sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/InlineCTE.scala

```diff
@@ -71,53 +71,64 @@ case class InlineCTE(
    * @param plan The plan to collect the CTEs from
    * @param cteMap A mutable map that accumulates the CTEs and their reference information by CTE
    *               ids.
-   * @param collectCTERefs A function to collect CTE references so that the caller side can do some
-   *                       bookkeeping work.
+   * @param outerCTEId While collecting the map we use this optional CTE id to identify the
+   *                   current outer CTE.
    */
   private def buildCTEMap(
       plan: LogicalPlan,
       cteMap: mutable.Map[Long, CTEReferenceInfo],
-      collectCTERefs: CTERelationRef => Unit = _ => ()): Unit = {
+      outerCTEId: Option[Long] = None): Unit = {
     plan match {
       case WithCTE(child, cteDefs) =>
-        cteDefs.foreach { cteDef =>
-          cteMap(cteDef.id) = CTEReferenceInfo(
-            cteDef = cteDef,
-            refCount = 0,
-            outgoingRefs = mutable.Map.empty.withDefaultValue(0),
-            shouldInline = true
-          )
-        }
-        cteDefs.foreach { cteDef =>
-          buildCTEMap(cteDef, cteMap, ref => {
-            // A CTE relation can references CTE relations defined before it in the same `WithCTE`.
-            // Here we update the out-going-ref-count for it, in case this CTE relation is not
-            // referenced at all and can be optimized out, and we need to decrease the ref counts
-            // for CTE relations that are referenced by it.
-            if (cteDefs.exists(_.id == ref.cteId)) {
-              cteMap(cteDef.id).increaseOutgoingRefCount(ref.cteId, 1)
-            }
-            // Similarly, a CTE relation can reference CTE relations defined in the outer `WithCTE`.
-            // Here we call the `collectCTERefs` function so that the outer CTE can also update the
-            // out-going-ref-count if needed.
-            collectCTERefs(ref)
-          })
+        val isDuplicated = cteDefs.forall(cteDef => cteMap.contains(cteDef.id))
+        if (isDuplicated) {
+          // If we have seen this `WithCTE` node then it must be self-contained so we can clear
+          // the references from containers to the definitions, and we don't need to process it
+          // again
+
+          cteDefs.foreach { cteDef =>
+            cteMap(cteDef.id).container.foreach(c => cteMap(c).outgoingRefs -= cteDef.id)
```

Comment:
note: Ideally we should do this for all self-contained CTEs, but only duplicated `WithCTE` has the problem, so this fix is fine.

Reply:
yeah, I mentioned that in the PR description

Reply:
A nested `WithCTE` is also duplicated, but it can legally reference outer CTE relations that are defined within this self-contained CTE.

Reply:
Can you give an example? But please note that we never visit any duplicates 2 times...

Reply:
oh you are right, when we hit the top-most duplicated `WithCTE`, we stop. Another trick we rely on to avoid issues...

Reply:
So what we really want to do is: the CTE refs inside the main query of this `WithCTE` should be excluded from container CTE relation's `outgoingRefs`, because this is a self-contained CTE. And because this is a self-contained CTE, the main query can only reference the CTE relations of this `WithCTE`, so we can just exclude all CTE relations of this `WithCTE`, instead of collecting the CTE refs from the main query.

Reply:
Yeah, technically this is what we do. We exclude all CTE definitions of the inner `WithCTE` from the container's `outgoingRefs`. (We don't collect and exclude only those defs that are in the inner's main query.)

Reply:
If it is confusuing that we store the container for each elements of `cteDefs` (as all have the same container), then we could store it for the first CTE def, and determine `isDuplicated` based on the presence of the first CTE def in the map. If that looks better...